.github/workflows/wheels.yml

```diff
@@ -0,0 +1,306 @@
+name: Release
+
+on:
+  push:
+    branches:
+      - main # just build the sdist & wheel, skip release
+    tags:
+      - "v*"
+  pull_request: # also build on PRs touching files that affect building sdist / wheels
+    paths:
+      - ".github/workflows/wheels.yml"
+      - "ci/**"
+      - "MANIFEST.in"
+      - "pyproject.toml"
+      - "setup.py"
+  workflow_dispatch:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  build-sdist:
+    name: Build fiona sdist
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Build source tarball
+        shell: bash
+        run: |
+          python3 -m venv venv && source venv/bin/activate
+          python -m pip install --upgrade pip
+          python -m pip install build
+          GDAL_VERSION=3.9.1 python -m build --sdist
+
+      - uses: actions/upload-artifact@v4
+        with:
+          name: fiona-sdist
+          path: ./dist/*.tar.gz
+          retention-days: 5
+          compression-level: 0
+
+  test-sdist:
+    name: Test sdist
+    needs: [build-sdist]
+    runs-on: ubuntu-latest
+    container:
+      image: "ghcr.io/osgeo/gdal:ubuntu-small-3.9.1"
+
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Download sdist from artifacts
+        uses: actions/download-artifact@v4
+        with:
+          name: fiona-sdist
+          path: wheelhouse
+
+      - name: Install pip
+        run: |
+          apt-get update && apt-get install -y python3-pip python3-dev python3-venv
+
+      - name: Build from sdist and install test dependencies
+        shell: bash
+        run: |
+          pwd
+          ls -l
+          python3 -m venv venv && source venv/bin/activate
+          python -m pip install --upgrade pip
+          python -m pip install --no-cache-dir wheelhouse/*.tar.gz
+          python -m pip install attrs pytest click mock boto3 packaging hypothesis wheel pytz fsspec aiohttp requests pyparsing shapely
+          python -m pip list
+
+      - name: Run tests
+        shell: bash
+        run: |
+          source venv/bin/activate
+          cd ..
+          mv Fiona/tests ./tests
+          python -m pytest -v -m "not wheel" tests
+
+  build-wheels-linux:
+    name: Build wheels on Linux
+    runs-on: "ubuntu-20.04"
```

Comment:
It might help to update to a newer Ubuntu release, which includes a later Python version on which the latest cibuildwheel runs.